### PR TITLE
Provide support for LZCNT and TZCNT in X86Ops.ins

### PR DESCRIPTION
**Pull request by VidushiGit**

### CODE REVIEW

**Comment by 0xdaryl** [comment]
@BradleyWood : please review

**Comment by 0xdaryl** [comment]
Thanks for the contribution.  Could you add encodings for the other register sizes as well please?  i.e., 16, 32, and 64-bit

**Comment by BradleyWood** [comment]
Can you add [binary encoding tests](https://github.com/eclipse/omr/blob/de6212d59d65524f189b3438b1c78ba74a524358/fvtest/compilerunittest/x/BinaryEncoder.cpp#L153) like the following?

`
INSTANTIATE_TEST_CASE_P(GeneralPurposeRegRegTest, XRegRegEncEncodingTest, ::testing::ValuesIn(*TRTest::MakeVector<std::tuple<TR::InstOpCode::Mnemonic, TR::RealRegister::RegNum, TR::RealRegister::RegNum, OMR::X86::Encoding, TRTest::BinaryInstruction>>(
    std::make_tuple(TR::InstOpCode::LZCNTRegReg,   TR::RealRegister::eax, TR::RealRegister::ecx, OMR::X86::Default, "<binary encoding>"),
    std::make_tuple(TR::InstOpCode::TZCNTRegReg,   TR::RealRegister::ecx, TR::RealRegister::eax, OMR::X86::Default, "<binary encoding>")
)));
`

**Comment by VidushiGit** [comment]
> Thanks for the contribution. Could you add encodings for the other register sizes as well please? i.e., 16, 32, and 64-bit

Yes, I will add support for the 16 and 32- bit also. Current implementation is only supporting 64 bits. Thanks

**Comment by VidushiGit** [comment]
Hi Bradley,
I have added the support for 16 and 32-bit registers for the instructions.

Also I have added the test cases for the instructions as suggested. Currently I have added EAX,ECX register pair(in the same order) for testing. Please let us know if we need some other test cases added.

File: fvtest/compilerunittest/x/BinaryEncoder.cpp [hunk 1]
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
     std::make_tuple(TR::InstOpCode::VPMOVB2MRegReg,    TR::RealRegister::k2, TR::RealRegister::xmm5,   OMR::X86::EVEX_L128,  "62f27e0829d5")
 )));
 
+INSTANTIATE_TEST_CASE_P(GeneralPurposeRegRegTest, XRegRegEncEncodingTest, ::testing::ValuesIn(*TRTest::MakeVector<std::tuple<TR::InstOpCode::Mnemonic, TR::RealRegister::RegNum, TR::RealRegister::RegNum, OMR::X86::Encoding, TRTest::BinaryInstruction>>(
+    std::make_tuple(TR::InstOpCode::LZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbdc1"),
+    std::make_tuple(TR::InstOpCode::LZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f34c0fbdc1"),
```
Comment (BradleyWood) [inline fvtest/compilerunittest/x/BinaryEncoder.cpp:700]: `f34c0fbdc1` equates to `lzcnt  r8,rcx` which does not match the specified registers

File: fvtest/compilerunittest/x/BinaryEncoder.cpp [hunk 2]
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
...
+
+    std::make_tuple(TR::InstOpCode::TZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbcc1"),
+    std::make_tuple(TR::InstOpCode::TZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f34c0fbcc1")
```
Comment (BradleyWood) [inline fvtest/compilerunittest/x/BinaryEncoder.cpp:703]: 'f34c0fbcc1' equates to `tzcnt  r8,rcx` which does not match the specified registers

**Review by BradleyWood — CHANGES_REQUESTED** [review]

**Comment by VidushiGit** [comment]
Corrected the test cases' binary encoding string

**Review by BradleyWood — APPROVED** [review]

File: compiler/x/codegen/X86Ops.ins [hunk 3]
```diff
@@ -5144,6 +5159,21 @@ INSTRUCTION(TEST8MemReg, test,
             PROPERTY0(IA32OpProp_ModifiesOverflowFlag | IA32OpProp_ModifiesSignFlag | IA32OpProp_ModifiesZeroFlag | IA32OpProp_ModifiesParityFlag | IA32OpProp_ModifiesCarryFlag | IA32OpProp_UsesTarget),
             PROPERTY1(IA32OpProp1_LongSource | IA32OpProp1_LongTarget | IA32OpProp1_FusableCompare),
             FEATURES(0)),
+INSTRUCTION(TZCNT2RegReg, tzcnt,
+            BINARY(VEX_L___, VEX_vNONE, PREFIX_F3, REX__, ESCAPE_0F__, 0xbc, 0, ModRM_RM__, Immediate_0),
```
Comment (0xdaryl) [inline compiler/x/codegen/X86Ops.ins:5163]: Same comment as above for the operand size prefix.

File: fvtest/compilerunittest/x/BinaryEncoder.cpp [hunk 4]
```diff
@@ -695,6 +695,14 @@ INSTANTIATE_TEST_CASE_P(AVX512MaskRegSimdEVEX512Test, XRegRegEncEncodingTest, ::
     std::make_tuple(TR::InstOpCode::VPMOVB2MRegReg,    TR::RealRegister::k2, TR::RealRegister::xmm5,   OMR::X86::EVEX_L128,  "62f27e0829d5")
 )));
 
+INSTANTIATE_TEST_CASE_P(GeneralPurposeRegRegTest, XRegRegEncEncodingTest, ::testing::ValuesIn(*TRTest::MakeVector<std::tuple<TR::InstOpCode::Mnemonic, TR::RealRegister::RegNum, TR::RealRegister::RegNum, OMR::X86::Encoding, TRTest::BinaryInstruction>>(
+    std::make_tuple(TR::InstOpCode::LZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbdc1"),
+    std::make_tuple(TR::InstOpCode::LZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f3480fbdc1"),
+
+    std::make_tuple(TR::InstOpCode::TZCNT4RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f30fbcc1"),
+    std::make_tuple(TR::InstOpCode::TZCNT8RegReg,   TR::RealRegister::eax,  TR::RealRegister::ecx, OMR::X86::Default, "f3480fbcc1")
```
Comment (0xdaryl) [inline fvtest/compilerunittest/x/BinaryEncoder.cpp:703]: Can you add tests for the 16-bit versions as well?

File: compiler/x/codegen/X86Ops.ins [hunk 5]
```diff
@@ -2545,6 +2545,21 @@ INSTRUCTION(LEA8RegMem, lea,
             PROPERTY0(IA32OpProp_ModifiesTarget),
             PROPERTY1(IA32OpProp1_LongSource | IA32OpProp1_LongTarget | IA32OpProp1_SourceIsMemRef),
             FEATURES(0)),
+INSTRUCTION(LZCNT2RegReg, lzcnt,
+            BINARY(VEX_L___, VEX_vNONE, PREFIX_F3, REX__, ESCAPE_0F__, 0xbd, 0, ModRM_RM__, Immediate_0),
```
[the remark this answers is not in the document]
Comment (BradleyWood) [reply compiler/x/codegen/X86Ops.ins:2549]: I had a look over the code, as it currently stands you cannot use both prefixes.

[the remark this answers is not in the document]
Comment (0xdaryl) [reply compiler/x/codegen/X86Ops.ins:2549]: PR #7020 was merged which provides the capability of specifying multiple prefixes.  @VidushiGit, you should pick up that change and use it here.

File: fvtest/compilerunittest/x/BinaryEncoder.cpp (hunk 4 above)
Comment (VidushiGit) [reply fvtest/compilerunittest/x/BinaryEncoder.cpp:705]: Added tests for 16-bit version of the instruction

File: compiler/x/codegen/X86Ops.ins (hunk 5 above)
[the remark this answers is not in the document]
Comment (VidushiGit) [reply compiler/x/codegen/X86Ops.ins]: @0xdaryl Pulled the latest OMR code changes and added the support/usage for multiple instruction prefixes

**Comment by BradleyWood** [comment]
jenkins build all

**Comment by BradleyWood** [comment]
MacOS failure is #6516

**Review by BradleyWood — APPROVED** [review]

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**